account-integrations/safe/test/hardhat/integration/SafeCompressionPlugin.test.ts

```diff
@@ -0,0 +1,198 @@
+
+    return {
+      factory: await ssf.connectOrDeploy(SafeProxyFactory__factory, []),
+      singleton: await ssf.connectOrDeploy(Safe__factory, []),
+      bundlerProvider,
+      provider,
+      userWallet,
+      entryPoints,
+    };
+  };
+
+  /**
+   * This test verifies a ERC4337 transaction succeeds when sent via a plugin
+   * The user operation deploys a Safe with the ERC4337 plugin and a handler
+   * and executes a transaction, thus verifying two things:
+   * 1. Deployment of the Safe with the ERC4337 plugin and handler is possible
+   * 2. Executing a transaction is possible
+   */
+  itif("should pass the ERC4337 validation", async () => {
+    const { singleton, provider, bundlerProvider, userWallet, entryPoints } =
+      await setupTests();
+
+    const ENTRYPOINT_ADDRESS = entryPoints[0];
+
+    const ssf = await SafeSingletonFactory.init(userWallet);
+
+    const safeCompressionFactory = await ssf.connectOrDeploy(
+      SafeCompressionFactory__factory,
+      [],
+    );
+
+    const feeData = await provider.getFeeData();
+    if (!feeData.maxFeePerGas || !feeData.maxPriorityFeePerGas) {
+      throw new Error(
+        "maxFeePerGas or maxPriorityFeePerGas is null or undefined",
+      );
+    }
+
+    const maxFeePerGas = `0x${feeData.maxFeePerGas.toString()}`;
+    const maxPriorityFeePerGas = `0x${feeData.maxPriorityFeePerGas.toString()}`;
+
+    const owner = ethers.Wallet.createRandom(provider);
+
+    await receiptOf(
+      userWallet.sendTransaction({
+        to: owner.address,
+        value: ethers.parseEther("100"),
+      }),
+    );
+
+    const addressRegistry = await ssf.connectOrDeploy(
+      AddressRegistry__factory,
+      [],
+    );
+
+    const fallbackDecompressor = await ssf.connectOrDeploy(
+      FallbackDecompressor__factory,
+      [await addressRegistry.getAddress()],
+    );
+
+    const createArgs = [
+      singleton,
+      ENTRYPOINT_ADDRESS,
+      await fallbackDecompressor.getAddress(),
+      owner.address,
+      0,
+    ] satisfies Parameters<typeof safeCompressionFactory.create.staticCall>;
+
+    const accountAddress = await safeCompressionFactory.create.staticCall(
+      ...createArgs,
+    );
+
+    await receiptOf(safeCompressionFactory.create(...createArgs));
+
+    const compressionAccount = SafeCompressionPlugin__factory.connect(
+      accountAddress,
+      userWallet,
+    );
+
+    const recipient = new ethers.Wallet(
+      "0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d",
+    );
+
+    const transferAmount = ethers.parseEther("1");
+
+    const compressedActions = await fallbackDecompressor.compress(
+      [
+        {
+          to: recipient.address,
+          value: transferAmount,
+          data: "0x",
+        },
+      ],
+      [],
+    );
+
+    const userOpCallData = compressionAccount.interface.encodeFunctionData(
+      "decompressAndPerform",
+      [compressedActions],
+    );
```

Comment:
Current implementation requires dApps (wallets or SDKs) to be compression-aware and wrap users' functions in `decompressAndPerform`. For broader use, this can eventually be implemented in an 'aggregator'?

Reply:
I think the wallet will always need to be compression-aware, similar to how it needs to be 4337-aware. Compression means that you actually sign a different message, so you can't give control over what message is signed to a third party. However, the same is true for using 4337 itself, so I'm not sure I understand your concern.